src/main/scala/firrtl/transforms/ConstantPropagation.scala

```diff
@@ -481,17 +481,23 @@ class ConstantPropagation extends Transform with ResolvedAnnotationPaths {
       propagated
     }
 
-    def backPropStmt(stmt: Statement): Statement = stmt map backPropExpr match {
-      case decl: IsDeclaration if swapMap.contains(decl.name) =>
-        val newName = swapMap(decl.name)
-        nPropagated += 1
-        decl match {
-          case node: DefNode => node.copy(name = newName)
-          case wire: DefWire => wire.copy(name = newName)
-          case reg: DefRegister => reg.copy(name = newName)
-          case other => throwInternalError()
-        }
-      case other => other map backPropStmt
+    def backPropStmt(stmt: Statement): Statement = stmt match {
+      case reg: DefRegister if (WrappedExpression.weq(reg.init, WRef(reg))) =>
+        // Self-init reset is an idiom for "no reset," and must be handled separately
+        val updatedName = swapMap.getOrElse(reg.name, reg.name)
+        reg.copy(name = updatedName, init = WRef(reg).copy(name = updatedName))
```

Comment:
I'm assuming it's intentional that you're not calling `map backPropExpr` on registers with self-init. Could you elaborate on why that is?

Reply:
Avoiding transforming the `init` field when it is a self-reference is basically the contribution of this PR. Otherwise, `ConstantPropagation` will transform something like this:
`reg r: UInt<1>, clk with: (reset => (rst, r))`

Into something like this for the case where `r` is constant zero:
`reg r: UInt<1>, clk with: (reset => (rst, UInt<1>(0)))`

Not calling `map backPropExpr` in those cases is the least invasive way to fix that problem.

The fact that it avoids calling `map backPropExpr` also impacts the other two expression fields: `clock` and `reset`. Since a self-reset register is already analyzed as having "no reset," and since there is a dependency on `RemoveReset` for register const prop to work at all, there is no benefit to calling `backPropExpr` on the `reset` expression, though I suppose it couldn't hurt. I sincerely hope that we aren't const-propping `clock` :)

Reply:
Got it, thanks for the explanation